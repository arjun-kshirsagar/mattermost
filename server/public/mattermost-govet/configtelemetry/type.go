// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package configtelemetry

import (
	"go/ast"
	"go/token"
	"strings"
)

var ignoreComments = []string{
	"telemetry: none",
	"Deprecated: do not use",
}

// typeFieldMap will generate map of its all possible selectors and adds them as a key. The value
// is the position of the declaration in the type spec.
func typeFieldMap(file *ast.File, typeName string) (map[string]token.Pos, error) {

	var fieldsMap map[string]token.Pos

	ast.Inspect(file, func(n ast.Node) bool {
		decl, ok := n.(*ast.GenDecl)
		if !ok || decl.Tok != token.TYPE || len(decl.Specs) < 1 {
			return true
		}

		spec, ok := decl.Specs[0].(*ast.TypeSpec)
		if !ok || spec.Name.Name != typeName {
			return true
		}

		typ, ok := spec.Type.(*ast.StructType)
		if !ok {
			return true
		}

		fieldsMap = fields(typ)

		return false
	})

	return fieldsMap, nil
}

func fields(str *ast.StructType) map[string]token.Pos {
	structFields := make(map[string]token.Pos)

	for _, field := range str.Fields.List {
		if field.Comment != nil && ignored(field.Comment) {
			continue
		}

		expr := field.Type
		if ptr, ok := field.Type.(*ast.StarExpr); ok {
			expr = ptr.X
		}

		if len(field.Names) != 1 {
			panic("unhandled struct field definition")
		}

		key := field.Names[0].Name

		v, ok := expr.(*ast.Ident)
		if !ok || v.Obj == nil || v.Obj.Decl == nil {
			structFields[key] = field.Pos()
			continue
		}

		typeDecl, ok := v.Obj.Decl.(*ast.TypeSpec)
		if !ok {
			continue
		}

		typ, ok := typeDecl.Type.(*ast.StructType)
		if !ok {
			continue
		}

		for k, v := range fields(typ) {
			structFields[key+"."+k] = v
		}
	}

	return structFields
}

func ignored(comments *ast.CommentGroup) bool {
	for _, comment := range comments.List {
		for _, ignore := range ignoreComments {
			if strings.Contains(comment.Text, ignore) {
				return true
			}
		}
	}
	return false
}
