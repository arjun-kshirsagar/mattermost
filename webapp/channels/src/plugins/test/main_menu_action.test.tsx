// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ComponentProps} from 'react';
import React from 'react';

import type {UserProfile} from '@mattermost/types/users';

import MainMenu from 'components/main_menu/main_menu';

import {renderWithContext, screen, userEvent} from 'tests/react_testing_utils';

describe('plugins/MainMenuActions', () => {
    const pluginAction = jest.fn();

    const requiredProps: ComponentProps<typeof MainMenu> = {
        teamId: 'someteamid',
        teamName: 'somename',
        currentUser: {id: 'someuserid', roles: 'system_user'} as UserProfile,
        enableCommands: true,
        enableIncomingWebhooks: true,
        enableOutgoingWebhooks: true,
        enableOAuthServiceProvider: true,
        canManageSystemBots: true,
        pluginMenuItems: [{
            id: 'someplugin',
            pluginId: 'test',
            text: 'some plugin text',
            action: pluginAction,
            mobileIcon: <i className={'fa fa-anchor'}/>,
        }],
        canManageIntegrations: true,
        moreTeamsToJoin: true,
        guestAccessEnabled: true,
        teamIsGroupConstrained: true,
        actions: {
            openModal: jest.fn(),
            showMentions: jest.fn(),
            showFlaggedPosts: jest.fn(),
            closeRightHandSide: jest.fn(),
            closeRhsMenu: jest.fn(),
        },
        isCloud: false,
        isStarterFree: false,
        canInviteTeamMember: false,
        isFreeTrial: false,
        usageDeltaTeams: -1,
        mobile: false,
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should render correctly in web view', () => {
        const {container} = renderWithContext(
            <MainMenu
                {...requiredProps}
            />,
        );

        expect(screen.getByText('some plugin text')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    test('should render correctly in mobile view with some plugin and ability to click plugin', () => {
        const props = {
            ...requiredProps,
            mobile: true,
        };

        const {container} = renderWithContext(
            <MainMenu
                {...props}
            />,
        );

        expect(screen.getByText('some plugin text')).toBeInTheDocument();
        expect(container).toMatchSnapshot();

        userEvent.click(screen.getByText('some plugin text'));
        expect(pluginAction).toHaveBeenCalled();
    });
});

