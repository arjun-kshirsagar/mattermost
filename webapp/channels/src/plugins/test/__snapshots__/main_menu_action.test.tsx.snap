// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/MainMenuActions should render correctly in mobile view with some plugin and ability to click plugin 1`] = `
<div>
  <div
    aria-label="main menu"
    class="a11y__popup Menu"
  >
    <ul
      class="Menu__content dropdown-menu"
      role="menu"
    >
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="recentMentions"
        role="menuitem"
      >
        <button
          class="style--none"
          data-testid="recentMentions-button"
          id="recentMentions-button"
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="mentions"
              >
                @
              </i>
            </span>
            <div
              class="text"
            >
              Recent Mentions
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuItem MenuItem--with-icon"
        id="flaggedPosts"
        role="menuitem"
      >
        <button
          class="style--none"
          data-testid="flaggedPosts-button"
          id="flaggedPosts-button"
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-bookmark"
              />
            </span>
            <div
              class="text"
            >
              Saved messages
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="profileSettings"
        role="menuitem"
      >
        <button
          aria-label="Profile"
          class="style--none "
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-user"
              />
            </span>
            <div
              class="text"
            >
              Profile
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuItem MenuItem--with-icon"
        id="accountSettings"
        role="menuitem"
      >
        <button
          aria-label="Settings"
          class="style--none "
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-cog"
              />
            </span>
            <div
              class="text"
            >
              Settings
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="viewMembers"
        role="menuitem"
      >
        <button
          aria-label="View Members"
          class="style--none "
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-users"
              />
            </span>
            <div
              class="text"
            >
              View Members
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="joinTeam"
        role="menuitem"
      >
        <a
          class=""
          href="/select_team"
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-plus-square"
              />
            </span>
            <div
              class="text"
            >
              Join Another Team
            </div>
          </span>
        </a>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="someplugin_pluginmenuitem"
        role="menuitem"
      >
        <button
          class="style--none"
          data-testid="someplugin_pluginmenuitem-button"
          id="someplugin_pluginmenuitem-button"
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-anchor"
              />
            </span>
            <div
              class="text"
            >
              some plugin text
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="nativeAppLink"
        role="menuitem"
      >
        <a
          href=""
          location="menu_item_external_link"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-mobile"
              />
            </span>
            <div
              class="text"
            >
              Download Apps
            </div>
          </span>
        </a>
      </li>
      <li
        class="MenuItem MenuItem--with-icon"
        id="about"
        role="menuitem"
      >
        <button
          aria-label="About Mattermost"
          class="style--none "
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-info"
              />
            </span>
            <div
              class="text"
            >
              About Mattermost
            </div>
          </span>
        </button>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuItem MenuItem--with-icon"
        id="logout"
        role="menuitem"
      >
        <button
          class="style--none"
          data-testid="logout-button"
          id="logout-button"
        >
          <span
            class="MenuItem__primary-text"
          >
            <span
              class="icon"
            >
              <i
                class="fa fa-sign-out"
              />
            </span>
            <div
              class="text"
            >
              Log Out
            </div>
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`plugins/MainMenuActions should render correctly in web view 1`] = `
<div>
  <div
    aria-label="team menu"
    class="a11y__popup Menu"
  >
    <ul
      class="Menu__content dropdown-menu"
      role="menu"
    >
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <li
        class="MenuItem"
        id="viewMembers"
        role="menuitem"
      >
        <button
          aria-label="View Members"
          class="style--none "
        >
          <span
            class="MenuItem__primary-text"
          >
            View Members
          </span>
        </button>
      </li>
      <li
        class="MenuItem"
        id="joinTeam"
        role="menuitem"
      >
        <a
          class=""
          href="/select_team"
        >
          <span
            class="MenuItem__primary-text"
          >
            Join Another Team
          </span>
        </a>
      </li>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: none;"
      />
      <div
        class="MainMenu_dropdown-link"
      >
        <div
          class="LearnAboutTeamsLink"
        >
          <a
            href="https://mattermost.com/pl/mattermost-academy-team-training?utm_source=mattermost&utm_medium=in-product&utm_content=learn_about_teams&uid=&sid="
            location="learn_about_teams"
            rel="noopener noreferrer"
            target="_blank"
          >
            <i
              class="icon icon-lightbulb-outline"
            />
            <span>
              Learn about teams
            </span>
          </a>
        </div>
      </div>
      <li
        class="MenuGroup menu-divider"
        role="separator"
        style="display: block;"
      />
      <li
        class="MenuItem"
        id="someplugin_pluginmenuitem"
        role="menuitem"
      >
        <button
          class="style--none"
          data-testid="someplugin_pluginmenuitem-button"
          id="someplugin_pluginmenuitem-button"
        >
          <span
            class="MenuItem__primary-text"
          >
            some plugin text
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
